awx/main/models/unified_jobs.py

```diff
@@ -357,18 +366,48 @@ def create_unified_job(self, **kwargs):
         if hasattr(self, 'survey_spec') and getattr(self, 'survey_enabled', False):
             password_list = self.survey_password_variables()
             hide_password_dict = getattr(unified_job, 'survey_passwords', {})
+            hide_password_dict.update(original_passwords)
             for password in password_list:
                 hide_password_dict[password] = REPLACE_STR
             unified_job.survey_passwords = hide_password_dict
 
         unified_job.save()
 
-        # Labels and extra credentials copied here
+        # Labels and credentials copied here
+        if kwargs.get('credentials'):
+            Credential = UnifiedJob._meta.get_field('credentials').related_model
+            cred_dict = Credential.unique_dict(self.credentials.all())
+            prompted_dict = Credential.unique_dict(kwargs['credentials'])
+            # combine prompted credentials with JT
+            cred_dict.update(prompted_dict)
+            kwargs['credentials'] = [cred for cred in cred_dict.values()]
+
         from awx.main.signals import disable_activity_stream
         with disable_activity_stream():
             copy_m2m_relationships(self, unified_job, fields, kwargs=kwargs)
+
+        if 'extra_vars' in kwargs:
+            unified_job.handle_extra_data(kwargs['extra_vars'])
+
+        if not getattr(self, '_deprecated_credential_launch', False):
+            # Create record of provided prompts for relaunch and rescheduling
+            unified_job.create_config_from_prompts(kwargs)
+
         return unified_job
 
+    @cached_subclassproperty
```

Comment:
Is calculating this really so expensive that we need to cache it?

Reply:
I agree caching is unimportant here. But the problem is that I still prefer something that can act as a class property, and it looks like it has to be custom. I don't know how to get that without defining a descriptor. I could change it to `subclassproperty` (and considered doing this), but that still leaves almost all of the added complexity.

Reply:
You can also accomplish something similar by defining a metaclass with a `@property`:

```diff
diff --git a/awx/main/models/unified_jobs.py b/awx/main/models/unified_jobs.py
index d19af29..48033a6 100644
--- a/awx/main/models/unified_jobs.py
+++ b/awx/main/models/unified_jobs.py
@@ -36,8 +36,7 @@ from awx.main.models.mixins import ResourceMixin, TaskManagerUnifiedJobMixin
 from awx.main.utils import (
     decrypt_field, _inventory_updates,
     copy_model_by_class, copy_m2m_relationships,
-    get_type_for_model, parse_yaml_or_json,
-    cached_subclassproperty
+    get_type_for_model, parse_yaml_or_json
 )
 from awx.main.redact import UriCleaner, REPLACE_STR
 from awx.main.consumers import emit_channel_notification
@@ -51,11 +50,29 @@ CAN_CANCEL = ('new', 'pending', 'waiting', 'running')
 ACTIVE_STATES = CAN_CANCEL
 
 
+class UnifiedJobTemplateMeta(type(PolymorphicModel), type(CommonModelNameNotUnique)):
+
+    @property
+    def ask_mapping(cls):
+        mapping = {}
+        for field in cls._meta.fields:
+            if not isinstance(field, AskForField):
+                continue
+            if field.allows_field == '__default__':
+                allows_field = field.name[len('ask_'):-len('_on_launch')]
+            else:
+                allows_field = field.allows_field
+            mapping[allows_field] = field.name
+        return mapping
+
+
 class UnifiedJobTemplate(PolymorphicModel, CommonModelNameNotUnique, NotificationFieldsModel):
     '''
     Concrete base class for unified job templates.
     '''
 
+    __metaclass__ = UnifiedJobTemplateMeta
+
     # status inherits from related jobs. Thus, status must be able to be set to any status that a job status is settable to.
     JOB_STATUS_CHOICES = [
         ('new', _('New')),                  # Job has been created, but not started.
@@ -395,19 +412,6 @@ class UnifiedJobTemplate(PolymorphicModel, CommonModelNameNotUnique, Notificatio
 
         return unified_job
 
-    @cached_subclassproperty
-    def ask_mapping(cls):
-        mapping = {}
-        for field in cls._meta.fields:
-            if not isinstance(field, AskForField):
-                continue
-            if field.allows_field == '__default__':
-                allows_field = field.name[len('ask_'):-len('_on_launch')]
-            else:
-                allows_field = field.allows_field
-            mapping[allows_field] = field.name
-        return mapping
-
     @classmethod
     def _get_unified_jt_copy_names(cls):
         return cls._get_unified_job_field_names()
diff --git a/awx/main/utils/common.py b/awx/main/utils/common.py
index cdb84b2..e1b414e 100644
--- a/awx/main/utils/common.py
+++ b/awx/main/utils/common.py
@@ -47,7 +47,7 @@ __all__ = ['get_object_or_400', 'get_object_or_403', 'camelcase_to_underscore',
            'extract_ansible_vars', 'get_search_fields', 'get_system_task_capacity',
            'wrap_args_with_proot', 'build_proot_temp_dir', 'check_proot_installed', 'model_to_dict',
            'model_instance_diff', 'timestamp_apiformat', 'parse_yaml_or_json', 'RequireDebugTrueOrTest',
-           'has_model_field_prefetched', 'set_environ', 'IllegalArgumentError', 'cached_subclassproperty',]
+           'has_model_field_prefetched', 'set_environ', 'IllegalArgumentError']
 
 
 def get_object_or_400(klass, *args, **kwargs):
@@ -935,17 +935,3 @@ def has_model_field_prefetched(model_obj, field_name):
     # NOTE: Update this function if django internal implementation changes.
     return getattr(getattr(model_obj, field_name, None),
                    'prefetch_cache_name', '') in getattr(model_obj, '_prefetched_objects_cache', {})
-
-
-class cached_subclassproperty(object):
-    '''Caches property in subclasses'''
-
-    def __init__(self, method):
-        self.method = method
-        self.name = method.__name__
-
-    def __get__(self, instance, cls):
-        r = self.method(cls)
-        if self.name not in cls.__dict__:
-            setattr(cls, self.name, r)
-        return r
```

Reply:
What if I just made it a `classmethod`, and called it like `JobTemplate.ask_mapping()` everywhere?

Reply:
That's fine, too :), it just sounded like you _really_ wanted a property-like interface here.  A _much_ simpler approach is to just make this a method.